src/drain-service/drain_training_inferencing.py

```diff
@@ -0,0 +1,345 @@
+# Standard Library
+import asyncio
+import json
+import logging
+import math
+import os
+import time
+from collections import deque
+
+# Third Party
+import numpy as np
+import pandas as pd
+from drain3.file_persistence import FilePersistence
+from drain3.template_miner import TemplateMiner
+from nats_wrapper import NatsWrapper
+
+pd.set_option("mode.chained_assignment", None)
+
+# Third Party
+from elasticsearch import AsyncElasticsearch
+from elasticsearch.helpers import BulkIndexError, async_streaming_bulk
+
+logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(name)s - %(message)s")
+persistence = FilePersistence("drain3_state.bin")
+template_miner = TemplateMiner(persistence)
+ES_ENDPOINT = os.environ["ES_ENDPOINT"]
+
+num_no_templates_change_tracking_queue = deque([], 50)
+num_templates_changed_tracking_queue = deque([], 50)
+num_clusters_created_tracking_queue = deque([], 50)
+num_total_clusters_tracking_queue = deque([], 50)
+
+
+async def consume_logs(nw, loop, incoming_logs_to_train_queue, logs_to_update_es):
+    if not nw.nc.is_connected:
+        await nw.connect(loop)
+        nw.add_signal_handler(loop)
+
+    async def subscribe_handler(msg):
+        subject = msg.subject
+        reply = msg.reply
+        payload_data = msg.data.decode()
+        payload_data = msg.data.decode()
+        await incoming_logs_to_train_queue.put(
+            pd.read_json(payload_data, dtype={"_id": object})
+        )
+
+    async def anomalies_subscription_handler(msg):
+        anomalies_data = msg.data.decode()
+        logging.info("got anomaly payload")
+        await logs_to_update_es.put(pd.read_json(anomalies_data, dtype={"_id": object}))
+
+    await nw.subscribe(
+        nats_subject="preprocessed_logs",
+        payload_queue=None,
+        subscribe_handler=subscribe_handler,
+    )
+    await nw.subscribe(
```

Comment:
would it be better to have a drain pod handling workload logs and another handling CP logs? If the load is heavy.

Reply:
Same pod is fine. I do think we should create a different DRAIN model for CP. Added this as a TODO